llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1084,6 +1086,15 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
     return;
   }
 
+  // (x && y) || (x && z) -> x && (y || z)
+  if (match(Def, m_c_BinaryOr(m_LogicalAnd(m_VPValue(X), m_VPValue(Y)),
+                              m_LogicalAnd(m_Deferred(X), m_VPValue(Z)))) &&
+      // Creating an extra recipe, so at least one arm needs to have one use.
+      (!Def->getOperand(0)->hasMoreThanOneUniqueUser() ||
+       !Def->getOperand(1)->hasMoreThanOneUniqueUser()))
```

Comment:
Why "more than one unique user"? Don't we just want to forbid the "no uses" case?

Reply:
If one of the arms has more than one use then it won't be dead after the transform. 

Given this transform introduces 2 recipes, the replacement should be at most 2 recipes. 

If both arms have more than one use then they'll still hang around, and we'll go from 3->4 recipes in total. 

So the guard that at least one arm will be dead is a heuristic to avoid the above scenario. We do it in quite a few places in InstCombine and DAGCombine IIRC

Reply:
Perhaps clarify that this is a heuristic in the comment, with a quick justification?